winsup/cygwin/include/cygwin/exit_process.h

```diff
@@ -0,0 +1,348 @@
+			   CREATE_NO_WINDOW | CREATE_NEW_PROCESS_GROUP,
+			   NULL, NULL, &si, &pi);
+  CloseHandle(h[1]);
+  if (!x)
+    {
+      CloseHandle(h[0]);
+      return NULL;
+    }
+
+  CloseHandle (pi.hThread);
+  CloseHandle (pi.hProcess);
+
+  char buffer[64];
+  DWORD offset = 0, count = 0;
+  while (offset < sizeof(buffer) - 1)
+    {
+      if (!ReadFile(h[0], buffer + offset, 1, &count, NULL))
+	{
+	  offset = 0;
+	  break;
+	}
+      if (!count || buffer[offset] == '\n')
+	break;
+      offset += count;
+    }
+  CloseHandle(h[0]);
+
+  buffer[offset] = '\0';
+
+  return (LPTHREAD_START_ROUTINE) strtoull(buffer, NULL, 16);
+}
+
+static int current_is_wow = -1;
+static int is_32_bit_os = -1;
+
+static BOOL get_wow(HANDLE process, BOOL& is_wow, int& bitness)
+{
+  if (is_32_bit_os == -1)
+    {
+      SYSTEM_INFO info;
+
+      GetNativeSystemInfo(&info);
+      if (info.wProcessorArchitecture == 0)
+	is_32_bit_os = 1;
+      else if (info.wProcessorArchitecture == 9)
+	is_32_bit_os = 0;
+      else
+	is_32_bit_os = -2;
+    }
+
+  if (current_is_wow == -1 &&
+      !IsWow64Process (GetCurrentProcess (), &current_is_wow))
+    current_is_wow = -2;
+
+  if (is_32_bit_os == -2 || current_is_wow == -2)
+    return FALSE;
+
+  if (!IsWow64Process (process, &is_wow))
+    return FALSE;
+
+  bitness = is_32_bit_os || is_wow ? 32 : 64;
+  return TRUE;
+}
+
+static LPTHREAD_START_ROUTINE
+get_exit_process_address_for_process(HANDLE process, int exit_code, DWORD pid)
+{
+  BOOL is_wow;
+  int bitness;
+  if (!get_wow (process, is_wow, bitness))
+    return NULL; /* could not determine */
+
+  if (is_wow == current_is_wow)
+    {
+      static LPTHREAD_START_ROUTINE exit_process_address;
+      if (!exit_process_address)
+	{
+	  HINSTANCE kernel32 = GetModuleHandle ("kernel32");
+	  exit_process_address = (LPTHREAD_START_ROUTINE)
+	    GetProcAddress (kernel32, "ExitProcess");
+	}
+      return exit_process_address;
+    }
+
+  /* Trying to terminate a 32-bit process from 64-bit or vice versa */
+  static LPTHREAD_START_ROUTINE exit_process_address;
+  if (!exit_process_address)
+    {
+      exit_process_address =
+	kill_via_console_helper (bitness, L"ExitProcess", exit_code, pid);
+    }
+  return exit_process_address;
+}
+
+static LPTHREAD_START_ROUTINE
+get_ctrl_routine_address_for_process_and_inject_remote_thread(HANDLE process, int exit_code, DWORD pid)
+{
+  BOOL is_wow;
+  int bitness;
+  if (!get_wow (process, is_wow, bitness))
+    return NULL; /* could not determine */
+
+  if (bitness == 64)
+    {
+      static LPTHREAD_START_ROUTINE ctrl_routine_address;
+      if (!ctrl_routine_address)
+	ctrl_routine_address =
+	  kill_via_console_helper (64, L"CtrlRoutine", exit_code, pid);
+      return ctrl_routine_address;
+    }
+
+  static LPTHREAD_START_ROUTINE ctrl_routine_address;
+  if (!ctrl_routine_address)
+    {
+      ctrl_routine_address =
+	kill_via_console_helper (32, L"CtrlRoutine", exit_code, pid);
+    }
```

Comment:
The difference between `bitness == 64` and `bitness == 32` now boils down to passing `bitness` as first parameter to `kill_via_console_helper()`. Therefore, we do not even need that `if (bitness == 64)` conditional at all anymore.